### ISSUE TITLE
Add functionality to `oak_sign` to sign challenge for user authentication

### ISSUE THREAD

**Issue by tiziano88** [project-oak/oak#1491]
Related to #1357 .

`oak_sign` may be extended so that it may be used offline to sign a challenge (from command line).

In the future, ideally this logic would instead be made part of a client library for connecting to Oak applications, which would generate key pairs, and use them to sign a challenge obtained from a specific Oak Runtime instance. For this, I think it may make sense to turn `oak_sign` into a library (i.e. separating, exposing and documenting the relevant functions) so that it may be used both by the `oak_sign` binary itself, as well as by a Rust client library.

cc @rbehjati @ipetr0v

**Comment by rbehjati** [comment]
I agree that this should be a library.

**Comment by ipetr0v** [comment]
> oak_sign may be extended so that it may be used offline to sign a challenge (from command line).

It's already used from a command line:
https://github.com/project-oak/oak/blob/605c64a90d279af254fef41be9b24609962edb30/scripts/oak_sign#L7

Also, agree about making it a library. It will also PEM related code:
https://github.com/project-oak/oak/blob/caa91c86f4f2b9d451d7a8bae9b56d82375f4b3e/oak_runtime/src/lib.rs#L1301-L1305

**Comment by tiziano88** [comment]
> > oak_sign may be extended so that it may be used offline to sign a challenge (from command line).
> 
> It's already used from a command line:
> https://github.com/project-oak/oak/blob/605c64a90d279af254fef41be9b24609962edb30/scripts/oak_sign#L7

Not for signing a challenge to be used for web auth though, right?

> Also, agree about making it a library. It will also PEM related code:
> https://github.com/project-oak/oak/blob/caa91c86f4f2b9d451d7a8bae9b56d82375f4b3e/oak_runtime/src/lib.rs#L1301-L1305

What about the PEM code (maybe you missed a word)?

**Comment by ipetr0v** [comment]
`It will also deduplicate PEM related code`, updated the original comment

**Comment by ipetr0v** [comment]
> Not for signing a challenge to be used for web auth though, right?

**Comment by tiziano88** [comment]
> > Not for signing a challenge to be used for web auth though, right?
> 
> Yes, but we will not be able to run a console command from a web browser.

Definitely, but I was thinking for initial experiments we would just get the challenge from the server, run `oak_sign` manually via command line, passing the challenge, and paste the signature back in the browser.

Anyways, since now we have an initial skeleton of a Chrome extension, it may actually make more sense to just implement this functionality in JS there, independent of `oak_sign`, but following the same exact logic and format if possible. Which means we should probably avoid PEM or even protobuf, otherwise we will need to have a corresponding library in JS for doing it.

Another option may be [JSON Web Signature](https://tools.ietf.org/html/rfc7515), but if it is anything like JWT (which has quite a few issues), then maybe it is not a great idea.

Basically we need a way of serializing together three fields:

- challenge (this is redundant, but it seems simpler to just include it)
- signature of the challenge using the private key
- public key

And the serialization should be achievable from JS.

I wonder whether we should just use a JSON object with three base64-encoded fields, similar to how we already have a JSON format for labels.

@rbehjati @ipetr0v thoughts?

**Comment by rbehjati** [comment]
Signing lib in JS plus JSON formatted challenge response sounds reasonable. The server could still support both JSON and protobuf (@juliettepretot previously mentioned [the need of using protobufs over HTTP](https://github.com/project-oak/oak/pull/1398#discussion_r481044555)). 

Does this mean that we can only allow JS clients for interaction over HTTP? Or could clients in other languages be used as well? In that case it may still make sense to make `oak_sign` a library that could be used in rust clients.

**Comment by tiziano88** [comment]
Not sure I see what the need of using protobufs over HTTP is for authentication specifically, which is not part of the business logic of the app, so not something that application developers would get to choose. I think we should pick one format and use it everywhere, and JSON seems the one that would be easiest to support on all platforms. In particular, if we decide to go that way, we may as well use the same format in `oak_sign`, for consistency.

**Comment by tiziano88** [comment]
In fact, I think that even for labels, we should perhaps just switch to JSON everywhere for simplicity, but that's a separate discussion (and not that urgent anyways).

**Comment by ipetr0v** [comment]
> run oak_sign manually via command line, passing the challenge, and paste the signature back in the browser.

Does Chrome allow access to the file system from an extension, without running Chrome with special flags?

> I wonder whether we should just use a JSON object with three base64-encoded fields, similar to how we already have a JSON format for labels.

I think it's the best option for us.

Also, do we need to send a JSON signature in every request?

**Comment by tiziano88** [comment]
> > run oak_sign manually via command line, passing the challenge, and paste the signature back in the browser.
> 
> Does Chrome allow access to the file system from an extension, without running Chrome with special flags?

https://developer.chrome.com/apps/fileSystem looks like it may do that, but what's your thought? Note that for now we don't even need persistent identities, so we can generate a key pair in JS directly.

In the future, if / when we want to support persistent identities, perhaps we could load the private key from disk, but that has its own set of issues, so I am not sure it is a great idea.

> 
> > I wonder whether we should just use a JSON object with three base64-encoded fields, similar to how we already have a JSON format for labels.
> 
> I think it's the best option for us.
> 
> Also, do we need to send a JSON signature in every request?

Yes, for the time being we will attach it as a header to every request. In the future we may consider keeping a connection open, and doing a single authentication per connection, then sending multiple requests on the same connection, but that's not going to be the case for a while.
